consensus/view_change_construct.go

```diff
@@ -0,0 +1,415 @@
+	// Set the bitmap indicating that this validator signed.
+	vc.viewIDBitmap[recvMsg.ViewID].SetKey(senderKey.Bytes, true)
+
+	return M2, nil, nil
+}
+
+// InitPayload do it once when validator received view change message
+// TODO: if I start the view change message, I should init my payload already
+func (vc *ViewChange) InitPayload(
+	fbftlog *FBFTLog,
+	viewID uint64,
+	blockNum uint64,
+	myPubKey string,
+	privKeys multibls.PrivateKeys,
+) error {
+	vc.vcLock.Lock()
+	defer vc.vcLock.Unlock()
+
+	// add my own M2 type message signature and bitmap
+	_, ok := vc.nilSigs[viewID][myPubKey]
+
+	if !ok {
+		vc.getLogger().Info().Uint64("viewID", viewID).Uint64("blockNum", blockNum).Msg("[InitPayload] add my M2 (NIL) type messaage")
+		for i, key := range privKeys {
+			if err := vc.nilBitmap[viewID].SetKey(key.Pub.Bytes, true); err != nil {
+				vc.getLogger().Warn().Int("index", i).Msg("[InitPayload] nilBitmap setkey failed")
+				continue
+			}
+			vc.nilSigs[viewID][key.Pub.Bytes.Hex()] = key.Pri.SignHash(NIL)
+		}
+	}
+
+	// add my own M3 type message signature and bitmap
+	_, ok = vc.viewIDSigs[viewID][myPubKey]
+
+	if !ok {
+		viewIDBytes := make([]byte, 8)
+		binary.LittleEndian.PutUint64(viewIDBytes, viewID)
+		vc.getLogger().Info().Uint64("viewID", viewID).Uint64("blockNum", blockNum).Msg("[InitPayload] add my M3 (ViewID) type messaage")
+		for i, key := range privKeys {
+			if err := vc.viewIDBitmap[viewID].SetKey(key.Pub.Bytes, true); err != nil {
+				vc.getLogger().Warn().Int("index", i).Msg("[InitPayload] viewIDBitmap setkey failed")
+				continue
+			}
+			vc.viewIDSigs[viewID][key.Pub.Bytes.Hex()] = key.Pri.SignHash(viewIDBytes)
+		}
+	}
+
+	// add my own M1 type message signature and bitmap
+	_, ok = vc.bhpSigs[viewID][myPubKey]
+
+	if !ok {
+		preparedMsgs := fbftlog.GetMessagesByTypeSeq(msg_pb.MessageType_PREPARED, blockNum)
+		preparedMsg := fbftlog.FindMessageByMaxViewID(preparedMsgs)
+		if preparedMsg != nil {
+			if preparedBlock := fbftlog.GetBlockByHash(preparedMsg.BlockHash); preparedBlock != nil {
+				if err := vc.blockVerifier(preparedBlock); err != nil {
+					return err
+				}
+				vc.getLogger().Info().Uint64("viewID", viewID).Uint64("blockNum", blockNum).Msg("[InitPayload] add my M1 (prepared) type messaage")
+				msgToSign := append(preparedMsg.BlockHash[:], preparedMsg.Payload...)
+				for i, key := range privKeys {
+					if err := vc.bhpBitmap[viewID].SetKey(key.Pub.Bytes, true); err != nil {
+						vc.getLogger().Warn().Int("index", i).Msg("[InitPayload] bhpBitmap setkey failed")
+						continue
+					}
+					vc.bhpSigs[viewID][key.Pub.Bytes.Hex()] = key.Pri.SignHash(msgToSign)
+				}
+				// if m1Payload is empty, we just add one
+				if vc.IsM1PayloadEmpty() {
+					vc.m1Payload = append(preparedMsg.BlockHash[:], preparedMsg.Payload...)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// IsM1PayloadEmpty returns true if m1Payload is not set
+func (vc *ViewChange) IsM1PayloadEmpty() bool {
```

Comment:
I see this function is used in both locked and unlocked version. Is is better to write a private version like: 
```
func (vc *ViewChange) IsM1PayloadEmpty() bool {
        vc.vcLock.RLock()
        defer vc.vcLock.RUnlock()
	return vc.isM1PayloadEmpty()
}

func (vc *ViewChange) isM1PayloadEmpty() bool {
	return len(vc.m1Payload) == 0
}
```

And private version can be used in ViewChange public methods.

Reply:
let me check, better protected in rlock.